Initial design description of processing system (#36)
Capstone-Skynet/Capstone-Skynet.github.io: docs/design/3_1_2_platform_management.tex

The platform management board manages all \textit{computing} I/O required by the multirotor: interfacing with the camera, programmable logic board (hosting the machine learning implementation), and the base-station. It is \textit{not} for use for flight control of the multirotor -- such systems are described in \ref{avionics}.

A decomposition of the system is seen in Figure XX

\begin{figure}\label{rpi}
\centering
\includegraphics[width=10cm]{img/RaspberryPi_Model_4B.png}
\caption{Layout of Connectors and Main ICs on Raspberry Pi 4}
\end{figure}

\textbf{Device Selection}

The selected device to implement the platform management system is the Raspberry Pi 4 (Model B). As seen in Figure \ref{rpi}, the Raspberry Pi 4 is a small single-board computer which hosts an ARM Cortex A72 CPU, a CSI camera interface, 40 serial pins, WiFi/Bluetooth, and a (1 gigabit) ethernet port. The Raspberry Pi 4 was selected over other single-computer platforms (such as Arduino/Beagleboard) as it:
\begin{itemize}

> \item is relatively inexpensive (\$45 in November 2019),

Reviewer: Use CAN$ notation for Canadian dollars.

\item has a large support community and a wealth of driver/library support,
\item supports a wide range of high throughput I/O standards (USB/WiFi/Ethernet/Serial, \textbf{F.CP.3}),

> \item provides an industry-standard camera interface (\textbf{F.CAM.2}),

Reviewer: ```suggestion
\item provides an industry-standard camera interface (\textbf{F.CAM.2}), and
```

\item supports extensible non-volatile storage (\textbf{F.CP.2})\\
\end{itemize} 

\textbf{Data Flow}

The overall data flow managed by the Raspberry Pi is as follows:
\begin{enumerate}

> \item Video is acquired through the camera CSI interface

Reviewer: Add periods at the end of itemized complete sentences.

\item Frame manipulation/correction is performed on the video, if required by the specific machine learning use-case
\item Manipulated video is decomposed and transmitted over Ethernet to the programmable logic board (which performs the machine learning)
\item Machine learning results are received via Ethernet
\item Video, ML results, and system status information are transmitted via WiFi to the base station
\item (Optional) Results and video are stored locally on non-volatile storage for post-flight analysis
\end{enumerate}

\textbf{Single-Board vs Two-Board Solution}

> The presence of on-board WiFi/Camera interfaces is the primary driving factor towards a two-board (Platform Management and Programmable) solution. While a single-board solution is possible, the development effort and cost of the computing platform would be considerably higher. Most programmable logic boards (including our chosen board, the Zedboard) require external expansion devices to interface with peripherals such as WiFi/cameras. These expansion devices, which connect to the programmable logic board via a serial connection, have poorly documented drivers and are of considerable expense -- ranging from \$25 USD for a basic WiFi module to \$200 USD for a compatible camera interface card. External expansion devices also require considerable glue logic resources on the programmable logic board, potentially risking non-compliance with constraint \textbf{C.EX.2}.

Reviewer: ```suggestion
The presence of on-board Wi-Fi/camera interfaces is the primary driving factor towards a two-board (PMB and PLB) solution. While a single-board solution is possible, the development effort and cost of the computing platform would be considerably higher. Most programmable logic boards (including our chosen board, the Zedboard) require external expansion devices to interface with peripherals such as Wi-Fi/cameras. These expansion devices, which connect to the programmable logic board via a serial connection, have poorly documented drivers and are of considerable expense --- ranging from US\$25 for a basic Wi-Fi module to US\$200 for a compatible camera interface card. External expansion devices also require considerable glue logic resources on the PLB, potentially risking non-compliance with the constraint \textbf{C.EX.2}.
```

Reviewer: Hmm. I guess the line is too long for git to pick up the diffs.

> An additional consideration in selecting the two-board design regards platform extensibility -- the client may desire to increase the Programmable Logic capabilities of the Computing Platform in the future, requiring a board swap. If a single-board solution is implemented, the client would be required to migrate \textit{all} hardware/software to the larger board, requiring them to devote considerable effort to debug driver issues. Additionally, as the external expansion devices are not necessarily compatible with all programmable logic boards -- adding constraints to the client's board selection or requiring the purchase of new interface cards. By offloading the interfacing work to a distinct \textit{platform management} board, the client can easily replace the PL board without considerable redevelopment (fulfilling constraint \textbf{C.EX.3}).

Reviewer: ```suggestion
An additional consideration in selecting the two-board design regards platform extensibility --- the client may desire to increase the Programmable Logic capabilities of the Computing Platform in the future, requiring a board swap. If a single-board solution is implemented, the client would be required to migrate \textit{all} hardware/software to the larger board, requiring them to devote considerable effort to debug driver issues. Additionally, as the external expansion devices are not necessarily compatible with all PLBs -- adding constraints to the client's board selection or requiring the purchase of new interface cards. By offloading the interfacing work to a distinct PMB, the client can easily replace the PLB without considerable redevelopment (fulfilling constraint \textbf{C.EX.3}).
```

\textbf{Interface with Programmable Logic Board}

> To interface with the programmable logic board, data packets are sent via a bi-directional TCP/IP-managed Ethernet connection. TCP/IP over Ethernet was selected in lieu of serial solutions (such as I2C) as it:

Reviewer: ```suggestion
To interface with the PLB, data packets are sent via a bi-directional TCP/IP-managed Ethernet connection. TCP/IP over Ethernet was selected in lieu of serial solutions (such as I2C) because:
```

\begin{itemize}

> \item is equally as ubiquitous on modern programmable logic boards,

Reviewer: ```suggestion
\item Ethernet is equally as ubiquitous on modern PLBs,
```

> \item has a considerably higher maximum throughput (1 Gbps vs 400Kbps) (\textbf{F.CP.3}),

Reviewer: ```suggestion
\item Ethernet has higher maximum throughput (1 Gbps on Ethernet vs. 400 Kbps on serial) (\textbf{F.CP.3}), and
```

> \item has more resilient error correction capabilities (\textbf{F.CP.1})

Reviewer: ```suggestion
\item Ethernet has more resilient error correction capabilities. (\textbf{F.CP.1})
```

\end{itemize} 

> The selection of Ethernet over serial precludes the use of the \textit{Zero W} variant of the Raspberry Pi, as the Zero W does not provide Ethernet support. While the Raspberry Pi 4 is considerably heavier (46g vs 9g), this difference is marginal when considering the total weight of the multirotor.

Reviewer: ```suggestion
The selection of Ethernet over serial precludes the use of the \textit{Zero W} variant of the Raspberry Pi, as the Zero W does not provide Ethernet support. While the Raspberry Pi 4 is considerably heavier (46 g for Raspberry Pi 4 vs. 9 g for Raspberry Pi Zero W), this difference is marginal when considering the total weight of the multirotor.
```
